tests/e2e_test.py

```diff
@@ -0,0 +1,118 @@
+            raise RuntimeError("Member does not exist")
+        return cast(int, member.id)
+
+
+def _get_auth_tokens() -> None:
+    IDP_URL_STAGING = "https://quantum-inspire-staging.eu.auth0.com"
+    QI2_DEFAULT_HOST = "https://staging.qi2.quantum-inspire.com"
+
+    E2E_USERNAME = os.getenv("E2E_USERNAME")
+    E2E_PASSWORD = os.getenv("E2E_PASSWORD")
+
+    payload = {
+        "grant_type": "password",
+        "client_id": "JOggYaBeGIkApEPIlQDZk8061Q8qHl4v",
+        "username": E2E_USERNAME,
+        "password": E2E_PASSWORD,
+        "scope": "api-access openid profile email offline_access",
+        "audience": "compute-job-manager",
+    }
+
+    headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+    url = f"{IDP_URL_STAGING}/oauth/token"
+
+    response = requests.post(url, data=payload, headers=headers)
+    response.raise_for_status()
+    token_info = response.json()
+    token_info["generated_at"] = time.time()
+    host = QI2_DEFAULT_HOST
+    member_id = asyncio.run(_fetch_team_member_id(host=host, access_token=token_info["access_token"]))
+    auth_settings = AuthSettings(
+        client_id="Yz7ni9PUAyT43eUASZfmc1yqI66QxLUJ",
+        code_challenge_method="S256",
+        code_verifyer_length=64,
+        well_known_endpoint=f"{IDP_URL_STAGING}/.well-known/openid-configuration",
+        tokens=None,
+        team_member_id=member_id,
+    )
+
+    dir_path = Path.home().joinpath(".quantuminspire")
+    dir_path.mkdir(parents=True, exist_ok=True)
+
+    ApiSettings(auths={host: auth_settings}, default_host=host).store_tokens(host=host, tokens=TokenInfo(**token_info))
+
+
+def _run_e2e_tests(backend_name: str) -> None:
+    from pennylane_quantuminspire2.qi_device import QI2Device
+
+    # Step 1: Select QML device
+    backend = QI2Device.get_backend(backend_name)
+    e2e_device = QI2Device(backend=backend)
+
+    # Step 2: Create a quantum circuit
+    @qml.qnode(e2e_device)
+    def my_quantum_circuit(circuit_params):  # type: ignore
+        qml.RX(circuit_params[0], wires=0)  # Apply an RX gate to qubit 0
+        qml.RY(circuit_params[1], wires=1)  # Apply an RY gate to qubit 1
+        qml.CNOT(wires=[0, 1])  # Apply a CNOT gate
+        return qml.expval(qml.PauliZ(0))  # Measure the expectation value of PauliZ on qubit 0
+
+    # Step 3: Initialize some parameters
+    params = np.array([0.1, 0.2], requires_grad=True)
+
+    # Step 4: Execute the circuit
+    result = my_quantum_circuit(params)
+    print(f"Params: {params}")
+    print(f"Result: {result}")
+
+    # Step 5: Perform optimization (optional)
+    # For example, use gradient descent to minimize the output
+    opt = qml.GradientDescentOptimizer(stepsize=0.1)
+    for i in range(3):
+        params = opt.step(my_quantum_circuit, params)
+    result = my_quantum_circuit(params)
+    print(f"Optimized params: {params}")
+    print(f"Optimized result: {result}")
+
+
+def main(name: str) -> None:
+    _get_auth_tokens()
```

Comment:
This function gets repeated here as well as in qiskit-quantum inspire. If there are changes in some parameters, we now have to change in 2 places. Lets discuss if we should just have it at a single location

Reply:
I agree this indeed is not ideal, would just have to think of the right place to put this. Maybe allow it to be imported from `qiskit-quantum-inspire`. Lets bring this up during refinement.

Reply:
Agreed, you can even say we are doing something like this in 3 places as we also do something similar for the platform E2E tests. Finding a proper place for this is kind of difficult though. We could put it into the `quantuminspire2` repository, but this is something that should only be used for tests and not made available as a feature to users. Maybe a custom shared github action?

Reply:
yes a custom github action would be a nice idea